commons/all/src/main/java/org/infinispan/commons/netty/BranchlessParser.java

```diff
@@ -0,0 +1,196 @@
+   }
+
+   private static long readRawVarInt72(ByteBuf buffer, long wholeOrMore) {
+      byte lastByte;
+      int skip = 9;
+      long msb;
+      // The number occupies all the available space.
+      // We have the special case for large negative numbers, where they occupy 9 bytes in varint representation.
+      // This way, we have to double-check whether it is a large positive or negative number.
+      // During this check, the number of bytes to consume and the MSB representation changes.
+      if (readableBytes(buffer) <= 8 || (lastByte = buffer.getByte(buffer.readerIndex() + 8)) < 0) {
+         if (readableBytes(buffer) >= 9) {
+            if ((lastByte = buffer.getByte(buffer.readerIndex() + 9)) < 0) {
+               buffer.resetReaderIndex();
+               return 0;
+            }
+
+            // A large negative number. We consume all 9 bytes and utilize the last 2 bytes to represent the MSB.
+            skip = 10;
+            msb = ((long) lastByte << 63) | (((long) buffer.getByte(8) & 0x7F) << 56);
+         } else {
+            buffer.resetReaderIndex();
+            return 0;
+         }
+      } else {
+         // A large positive number.
+         // We consume 8 bytes and use only the last byte for the MSB.
+         msb = (long) lastByte << 56;
+      }
+
+      buffer.skipBytes(skip);
+      return msb | readLong(wholeOrMore);
+   }
+
+   private static long readLong(long continuation) {
+      // We parse it as groups of bytes, first bytes 1, 3, 5, and 7.
+      // The second group is 2, 4, 6, and 8, which need a shift to right to compensate the gap.
+      continuation = (continuation & 0x007F007F007F007FL) | ((continuation & 0x7F007F007F007F00L) >> 1);
+
+      // Now we isolate the bits in sequence. We check 14 bits at a time.
+      // We shift it back to compensate the continuation bits. Since we group two bytes, we shift in increasing steps.
+      return (continuation & 0x3FFF) |
+            ((continuation & 0x3FFF0000) >> 2) |
+            ((continuation & 0x3FFF00000000L) >> 4) |
+            ((continuation & 0x3FFF000000000000L) >> 6);
+   }
+
+   private static int readRawVarint24(ByteBuf buffer) {
+      if (!buffer.isReadable())
+         return 0;
+
+      // Reaching this point, we have at most 3 bytes.
+      // It is either a smaller number or the buffer still haven't all the necessary bytes to conclude.
+      // The number might occupy 1, 2, or 3 bytes. This way, we have to read as much as we can to check.
+      int start = buffer.readerIndex();
+
+      // We read the first byte.
+      // If it is zero or positive, means we have read the complete value.
+      // Otherwise, the number occupies more bytes because the 8 bit is set, marking a continuation.
+      byte b = buffer.readByte();
+      if (b >= 0)
+         return b;
+
+      // The number has the continuation bit set, but we can't read any more bytes.
+      // We reset and return.
+      if (!buffer.isReadable()) {
```

Comment:
IIRC all of these calls to isReadable are superfluous. Also checking the index shouldn't be needed either. All you have to do is call `readByte`. Then the HintingByteBuf will throw a SIGNAL which will be propagated up the stack and it HintedReplayingDecoder will reset the position back to the last decoded position.